# Importing Libraries

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import scipy

In [2]:
# setting path
path = r'C:\Users\faisa\Desktop\Data analysis\Data Immersion\Achievement 6'

In [3]:
# importing data sets
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)
df_products = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_users = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'users.csv'), index_col = False)
df_inventory_items = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'inventory_items.csv'), index_col = False)
df_order_items = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'order_items.csv'), index_col = False)
df_events = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'events.csv'), index_col = False)
df_distribution_centers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'distribution_centers.csv'), index_col = False)

# Script

## Data Cleaning

### Data cleaning: Orders

In [4]:
df_orders.tail()

order_id  user_id   status gender                 created_at  \
125221    125196    99972  Shipped      M  2022-06-25 11:56:00+00:00   
125222    125199    99975  Shipped      M  2022-07-20 02:19:00+00:00   
125223    125208    99983  Shipped      M  2022-03-14 15:06:00+00:00   
125224    125210    99983  Shipped      M  2023-03-01 15:06:00+00:00   
125225    125224    99999  Shipped      M  2023-12-22 11:29:00+00:00   

       returned_at                 shipped_at delivered_at  num_of_item  
125221         NaN  2022-06-28 08:24:00+00:00          NaN            1  
125222         NaN  2022-07-22 11:08:00+00:00          NaN            4  
125223         NaN  2022-03-17 05:15:00+00:00          NaN            1  
125224         NaN  2023-03-02 10:50:00+00:00          NaN            2  
125225         NaN  2023-12-24 00:44:00+00:00          NaN            2

In [5]:
# Checking for value consistency
df_orders['num_of_item'].value_counts(ascending = False)

num_of_item
1    87712
2    24867
4     6372
3     6275
Name: count, dtype: int64

In [6]:
# checking for duplicates
df_orders[df_orders.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, status, gender, created_at, returned_at, shipped_at, delivered_at, num_of_item]
Index: []

In [5]:
df_orders.shape

(125226, 9)

In [7]:
df_products.shape

(29120, 9)

In [8]:
df_users.shape

(100000, 15)

In [9]:
df_inventory_items.shape

(490705, 12)

In [10]:
df_order_items.shape

(181759, 11)

In [11]:
df_events.shape

(2431963, 13)

In [12]:
df_distribution_centers.shape

(10, 4)

In [50]:
# checking variable tyoe
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125226 entries, 0 to 125225
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   order_id      125226 non-null  int64 
 1   user_id       125226 non-null  int64 
 2   status        125226 non-null  object
 3   gender        125226 non-null  object
 4   created_at    125226 non-null  object
 5   returned_at   12530 non-null   object
 6   shipped_at    81461 non-null   object
 7   delivered_at  43884 non-null   object
 8   num_of_item   125226 non-null  int64 
dtypes: int64(3), object(6)
memory usage: 8.6+ MB


In [7]:
# checking missing values
df_orders.isnull().sum()

order_id             0
user_id              0
status               0
gender               0
created_at           0
returned_at     112696
shipped_at       43765
delivered_at     81342
num_of_item          0
dtype: int64

### Data cleaning: Products

In [6]:
df_products.head()

id     cost     category  \
0  13842  2.51875  Accessories   
1  13928  2.33835  Accessories   
2  14115  4.87956  Accessories   
3  14157  4.64877  Accessories   
4  14273  6.50793  Accessories   

                                                name brand  retail_price  \
0   Low Profile Dyed Cotton Twill Cap - Navy W39S55D    MG          6.25   
1  Low Profile Dyed Cotton Twill Cap - Putty W39S55D    MG          5.95   
2       Enzyme Regular Solid Army Caps-Black W35S45D    MG         10.99   
3  Enzyme Regular Solid Army Caps-Olive W35S45D (...    MG         10.99   
4              Washed Canvas Ivy Cap - Black W11S64C    MG         15.99   

  department                               sku  distribution_center_id  
0      Women  EBD58B8A3F1D72F4206201DA62FB1204                       1  
1      Women  2EAC42424D12436BDD6A5B8A88480CC3                       1  
2      Women  EE364229B2791D1EF9355708EFF0BA34                       1  
3      Women  00BD13095D06C20B11A2993CA419D16B                       1  
4      Women  F531DC20FDE20B7ADF3A73F52B71D0AF                       1

In [175]:
# dropping sku column
df_products = df_products.drop (columns = ['sku'])

In [9]:
df_products['brand'].value_counts()

brand
Allegra K               1034
Calvin Klein             497
Carhartt                 388
Hanes                    308
Volcom                   292
                        ... 
Legendary Apparel          1
Nygård Collection         1
A Christmas Story          1
Pookies by DFDesigns       1
DeepPocket                 1
Name: count, Length: 2756, dtype: int64

In [10]:
# checking for duplicates
df_products[df_products.duplicated()]

Empty DataFrame
Columns: [id, cost, category, name, brand, retail_price, department, sku, distribution_center_id]
Index: []

In [42]:
df_products.isnull().sum()

id                         0
cost                       0
category                   0
name                       2
brand                     24
retail_price               0
department                 0
sku                        0
distribution_center_id     0
dtype: int64

In [44]:
# Create a map to pick the most frequent brand used according to name that is not null
map_brand = (
    df_products[df_products['brand'].notnull()]
    .groupby('name')['brand']
    .agg(lambda x: x.mode().iloc[0])
)

In [48]:
# Fill missing brand using the map_brand
df_products['brand'] = df_products.apply(
    lambda row: map_brand[row['name']] if pd.isnull(row['brand']) and row['name'] in map_brand else row['brand'],
    axis=1
)

In [49]:
df_products.isnull().sum()

id                         0
cost                       0
category                   0
name                       2
brand                     22
retail_price               0
department                 0
sku                        0
distribution_center_id     0
dtype: int64

In [116]:
# checking data type
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29120 entries, 0 to 29119
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      29120 non-null  int64  
 1   cost                    29120 non-null  float64
 2   category                29120 non-null  object 
 3   name                    29118 non-null  object 
 4   brand                   29096 non-null  object 
 5   retail_price            29120 non-null  float64
 6   department              29120 non-null  object 
 7   sku                     29120 non-null  object 
 8   distribution_center_id  29120 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 2.0+ MB


_Previous Missing values in 'brand' = 24
New Missing values in 'brand' = 22_

### Data cleaning: Users

In [11]:
df_users.head()

id   first_name last_name                           email  age gender  \
0    457      Timothy      Bush         timothybush@example.net   65      M   
1   6578    Elizabeth  Martinez   elizabethmartinez@example.com   34      F   
2  36280  Christopher   Mendoza  christophermendoza@example.net   13      M   
3  60193        Jimmy    Conner         jimmyconner@example.com   64      M   
4  64231      Natasha    Wilson       natashawilson@example.net   25      F   

  state                street_address postal_code        city country  \
0  Acre           87620 Johnson Hills   69917-400  Rio Branco  Brasil   
1  Acre             1705 Nielsen Land   69917-400  Rio Branco  Brasil   
2  Acre      125 Turner Isle Apt. 264   69917-400  Rio Branco  Brasil   
3  Acre     0966 Jose Branch Apt. 008   69917-400  Rio Branco  Brasil   
4  Acre  20798 Phillip Trail Apt. 392   69917-400  Rio Branco  Brasil   

   latitude  longitude traffic_source                 created_at  
0 -9.945568  -67.83561         Search  2022-07-19 13:51:00+00:00  
1 -9.945568  -67.83561         Search  2023-11-08 18:49:00+00:00  
2 -9.945568  -67.83561          Email  2019-08-24 06:10:00+00:00  
3 -9.945568  -67.83561         Search  2020-02-15 11:26:00+00:00  
4 -9.945568  -67.83561         Search  2020-03-13 06:45:00+00:00

In [12]:
df_users.describe()

id            age       latitude      longitude
count  100000.000000  100000.000000  100000.000000  100000.000000
mean    50000.500000      41.054790      28.315885      25.020268
std     28867.657797      17.043914      22.024165      89.989960
min         1.000000      12.000000     -43.007536    -158.164931
25%     25000.750000      26.000000      26.148665     -50.794605
50%     50000.500000      41.000000      35.219886       4.793600
75%     75000.250000      56.000000      40.707772     116.376103
max    100000.000000      70.000000      64.865194     153.560238

In [13]:
# checking for duplicates
df_users[df_users.duplicated()]

Empty DataFrame
Columns: [id, first_name, last_name, email, age, gender, state, street_address, postal_code, city, country, latitude, longitude, traffic_source, created_at]
Index: []

In [14]:
# checking missing values
df_users.isnull().sum()

id                  0
first_name          0
last_name           0
email               0
age                 0
gender              0
state               0
street_address      0
postal_code         0
city              958
country             0
latitude            0
longitude           0
traffic_source      0
created_at          0
dtype: int64

In [57]:
# Create a map to pick the most frequent city used according to postal code that is not null
map_user_city = (
    df_users[df_users['city'].notnull()]
    .drop_duplicates(subset=['postal_code'])  # In case of duplicates, keep one
    .set_index('postal_code')['city']
)

In [58]:
# Fill missing cities using the map_user_city
df_users['city'] = df_users.apply(
    lambda row: map_user_city[row['postal_code']] if pd.isnull(row['city']) and row['postal_code'] in map_user_city else row['city'],
    axis=1
)

In [59]:
# checking missing values
df_users.isnull().sum()

id                  0
first_name          0
last_name           0
email               0
age                 0
gender              0
state               0
street_address      0
postal_code         0
city              945
country             0
latitude            0
longitude           0
traffic_source      0
created_at          0
dtype: int64

_since there was no cities in the data frame with the same postal code. we move on to do it manually_

In [86]:
df_users[df_users['city'].str.contains('NY', case=False, na=False)].head(60)

id first_name  last_name                          email  age gender  \
2413   2583      Molly      Estes         mollyestes@example.net   22      F   
2414  11002     Ashley     Santos       ashleysantos@example.org   41      F   
2415  21935    Michael    Baldwin     michaelbaldwin@example.net   31      M   
2416  23176  Elizabeth    Stanley   elizabethstanley@example.org   61      F   
2417  25955     Brooke     Mccann       brookemccann@example.com   56      F   
2418  63116      Aaron     Rogers        aaronrogers@example.net   32      M   
2419  64569    Katrina     Fisher      katrinafisher@example.net   70      F   
2420  67200    Jessica  Rodriguez   jessicarodriguez@example.com   34      F   
2421  79569     Morgan      Ayala        morganayala@example.org   22      F   
2422  79862    Jeffrey      Henry       jeffreyhenry@example.com   69      M   
2423  86747      Renee      Meyer         reneemeyer@example.com   50      F   
2424  88120      Heidi    Nichols       heidinichols@example.com   66      F   
2425  93393       Mary    Hartman        maryhartman@example.com   35      F   
2719   8664        Jim     Sutton          jimsutton@example.org   23      M   
2720  14315       John     Holden         johnholden@example.org   24      M   
2721  15204     Steven      Nixon        stevennixon@example.net   25      M   
2722  18801    Gregory     Taylor      gregorytaylor@example.com   68      M   
2723  22859    Kenneth     Bryant      kennethbryant@example.com   63      M   
2724  24933    Cynthia     Watson      cynthiawatson@example.org   43      F   
2725  29232   Meredith  Whitehead  meredithwhitehead@example.com   14      F   
2726  43030     Andrew    Bradley      andrewbradley@example.org   23      M   
2727  44441      Shane     Walker        shanewalker@example.com   69      M   
2728  52045       Gary       Cruz           garycruz@example.org   23      M   
2729  54048     Jeremy    Barrett      jeremybarrett@example.org   69      M   
2730  55268     Cheryl       Kane         cherylkane@example.com   21      F   
2731  55698      Peter  Valentine     petervalentine@example.org   62      M   
2732  66014    Phillip     Taylor      philliptaylor@example.net   14      M   
2733  73473     Steven     Brooks       stevenbrooks@example.org   55      M   
2734  74171    Felicia   Thompson    feliciathompson@example.com   35      F   
2735  81971       Glen      White          glenwhite@example.net   15      M   
2736  84276    Matthew   Harrison    matthewharrison@example.net   56      M   
2737  89235    Shelley   Caldwell    shelleycaldwell@example.com   66      F   
2738  93837      Kevin    Morales       kevinmorales@example.net   24      M   
2739  94957      James    Edwards       jamesedwards@example.org   38      M   
2740  98141    Loretta     Garcia      lorettagarcia@example.org   33      F   
2976   6911   Courtney       Vang       courtneyvang@example.net   19      F   
2977   8295  Johnathan      Brown     johnathanbrown@example.net   66      M   
2978  44446    Gregory      Chung       gregorychung@example.net   59      M   
2979  58389      Casey        Ray           caseyray@example.com   21      M   
2980  64237    Jasmine    Mueller     jasminemueller@example.net   19      F   
3852  44678      Craig      Huynh         craighuynh@example.org   68      M   
3853  65553    Charles       Chen        charleschen@example.com   36      M   
3893   4675      Erica   Williams      ericawilliams@example.com   32      F   
3894  21315     Thomas     Nguyen       thomasnguyen@example.org   27      M   
3895  28250      April     Weaver        aprilweaver@example.org   53      F   
3896  65653    Jessica     Willis      jessicawillis@example.net   45      F   
3897  95329     Meghan     Mosley       meghanmosley@example.org   40      F   
8079   3693      James   Santiago      jamessantiago@example.net   36      M   
8080   5048    Michael      Chase       michaelchase@example.net   37      M   
8082   76

In [173]:
# Dropping email, street_address, latitude, longitude columns
df_users = df_users.drop (columns = ['email', 'street_address', 'latitude', 'longitude'])

In [117]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 15 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              100000 non-null  int64  
 1   first_name      100000 non-null  object 
 2   last_name       100000 non-null  object 
 3   email           100000 non-null  object 
 4   age             100000 non-null  int64  
 5   gender          100000 non-null  object 
 6   state           100000 non-null  object 
 7   street_address  100000 non-null  object 
 8   postal_code     100000 non-null  object 
 9   city            100000 non-null  object 
 10  country         100000 non-null  object 
 11  latitude        100000 non-null  float64
 12  longitude       100000 non-null  float64
 13  traffic_source  100000 non-null  object 
 14  created_at      100000 non-null  object 
dtypes: float64(2), int64(2), object(11)
memory usage: 11.4+ MB


In [91]:
# finding distinct postal codes that have missing cities
missing_city_postal = df_users[df_users['city'].isnull()]['postal_code'].dropna().unique()

In [92]:
missing_city_postal

array(['29140', '41130', '50016', '50059', '50191', '50197', '38297',
       '38300', '38434', '25191', '46988', '46394', '28524', '72025-065',
       '72110-600', '72016-190', '29142', '34476', '34482', '48044',
       '77389', '77379', '79706', '300384', '20171', '20112'],
      dtype=object)

In [93]:
# creating a data dictionary combining postal codes with cities
postcode_to_city = {
    '69980-000': 'Rio Branco',
    '57360-000': 'Maribondo',
    '68924-000': 'Laranjal do Jari',
    '69190-000': 'São Miguel do Guaporé',
    '69230-000': 'Seringueiras',
    '69200-000': 'Guajará-Mirim',
    '69735-000': 'Porto Walter',
    '69250-000': 'Pimenta Bueno',
    '69435-000': 'Jordão',
    '69280-000': 'Cacoal',
    '69800-000': 'Xapuri',
    '69460-000': 'Tarauacá',
    '69830-000': 'Epitaciolândia',
    '69600-000': 'Cruzeiro do Sul',
    '69880-000': 'Brasileia',
    '69630-000': 'Feijó',
    '48475-000': 'Conceição do Coité',
    '48790-000': 'Jaguaquara',
    '44330-000': 'Jacobina',
    '48760-000': 'Jequié',
    '47350-000': 'Irecê',
    '47240-000': 'Ibititá',
    '63540-000': 'Tauá',
    '63560-000': 'Tamboril',
    '63610-000': 'Campos Sales',
    '62150-000': 'Mauriti',
    '65530-000': 'Paraíso do Tocantins',
    '65690-000': 'Santa Helena',
    '65415-000': 'Buritirana',
    '65725-000': 'Matupá',
    '65278-000': 'São Bento',
    '65365-000': 'Formosa do Rio Preto',
    '65393-000': 'São Desidério',
    '65380-000': 'Correntina',
    '78850-000': 'Juara',
    '39260-000': 'Juiz de Fora',
    '84600-000': 'Palmas',
    '68617-000': 'Benevides',
    '68620-000': 'Castanhal',
    '68650-000': 'Bragança',
    '68660-000': 'Bragança',
    '68695-000': 'Vigia',
    '68540-000': 'Santarém',
    '68473-000': 'Tailândia',
    '68890-000': 'Marabá',
    '68300-000': 'Itaituba',
    '68330-000': 'Rurópolis',
    '68230-000': 'Alenquer',
    '68250-000': 'Altamira',
    '55920-000': 'Petrolina',
    '55730-000': 'Salgueiro',
    '56460-000': 'Ouricuri',
    '56180-000': 'Araripina',
    '56380-000': 'Cabrobó',
    '76980-000': 'Ji-Paraná',
    '76868-000': 'Porto Velho',
    '73900-000': 'Formosa',
    '75370-000': 'Anápolis',
    '10001': 'New York',
    '19803': 'Wilmington',
    '12065': 'Glens Falls',
    '13090': 'Syracuse',
    '15174': 'Pittsburgh',
    '17015': 'Harrisburg',
    '18302': 'Stroudsburg',
    '22406': 'Fredericksburg',
    '22407': 'Fredericksburg',
    '22408': 'Fredericksburg',
    '22556': 'Stafford',
    '22602': 'Winchester',
    '23188': 'Richmond',
    '23236': 'Richmond',
    '23693': 'Newport News',
    '27537': 'Cary',
    '28909': 'Bryson City',
    '28946': 'Sylva',
    '28947': 'Sylva',
    '30016': 'Conyers',
    '30028': 'Decatur',
    '30044': 'Lithonia',
    '30045': 'Lithonia',
    '30093': 'Norcross',
    '30102': 'Acworth',
    '30835': 'Augusta',
    '32940': 'Melbourne',
    '32826': 'Orlando',
    '32828': 'Orlando',
    '32092': 'Saint Augustine',
    '33437': 'Delray Beach',
    '33579': 'Plant City',
    '33900': 'Fort Myers',
    '33129': 'Miami',
    '88201': 'Las Cruces',
    '98926': 'Yakima',
    '70706': 'Baton Rouge',
    '631-830': 'St. Louis', 
    '631-820': 'St. Louis',
    '631-810': 'St. Louis',
    '630-850': 'Saint Charles',
    '630-857': 'Saint Charles',
    '630-040': 'Ballwin',
    '630-520': 'Ballwin',
    '631-840': 'St. Louis',
    '630-492': 'Ballwin',
    '630-490': 'Ballwin',
    '641-870': 'Kansas City',
    '630-500': 'Ballwin',
    '642-370': 'Kansas City',
    '641-465': 'Kansas City',
    '641-860': 'Kansas City',
    '642-315': 'Kansas City',
    '641-920': 'Kansas City',
    '29140': 'Summerton, South Carolina',
    '41130': 'Flatwoods, Kentucky',
    '50016': 'Bondurant, Iowa',
    '50059': 'Des Moines, Iowa',
    '50191': 'Woodward, Iowa',
    '50197': 'Waukee, Iowa',
    '38297': 'Tiptonville, Tennessee',
    '38300': 'Jackson, Tennessee',
    '38434': 'Columbia, Tennessee',
    '25191': 'Saint Albans, West Virginia',
    '46988': 'Wabash, Indiana',
    '46394': 'Valparaiso, Indiana',
    '28524': 'Beaufort, North Carolina',
    '72025-065': 'Conway, Arkansas',
    '72110-600': 'Cabot, Arkansas',
    '72016-190': 'Bald Knob, Arkansas',
    '29142': 'Sumter, South Carolina',
    '34476': 'Ocala, Florida',
    '34482': 'Ocala, Florida',
    '48044': 'Eastpointe, Michigan',
    '77389': 'The Woodlands, Texas',
    '77379': 'The Woodlands, Texas',
    '79706': 'Midland, Texas',
    '300384': 'Unknown', 
    '20171': 'Manassas, Virginia',
    '20112': 'Chantilly, Virginia'
}

In [94]:
# filling missing cities based on the dictionary created.
df_users['city'] = df_users.apply(
    lambda row: postcode_to_city.get(row['postal_code'], row['city']) if pd.isnull(row['city']) else row['city'],
    axis=1
)

In [95]:
df_users[df_users['city'].isnull()]

Empty DataFrame
Columns: [id, first_name, last_name, email, age, gender, state, street_address, postal_code, city, country, latitude, longitude, traffic_source, created_at]
Index: []

In [96]:
# checking missing values
df_users.isnull().sum()

id                0
first_name        0
last_name         0
email             0
age               0
gender            0
state             0
street_address    0
postal_code       0
city              0
country           0
latitude          0
longitude         0
traffic_source    0
created_at        0
dtype: int64

_Previous number of missing City: 945
Updated number of missing City: 0_

### Data cleaning: Inventory_items

In [15]:
df_inventory_items.head()

id  product_id                 created_at                    sold_at  \
0  67971       13844  2022-07-02 07:09:20+00:00  2022-07-24 06:33:20+00:00   
1  67972       13844  2023-12-20 03:28:00+00:00                        NaN   
2  67973       13844  2023-06-04 02:53:00+00:00                        NaN   
3  72863       13844  2021-10-16 22:58:52+00:00  2021-11-22 02:19:52+00:00   
4  72864       13844  2021-08-07 16:33:00+00:00                        NaN   

      cost product_category            product_name       product_brand  \
0  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   
1  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   
2  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   
3  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   
4  2.76804      Accessories  (ONE) 1 Satin Headband  Funny Girl Designs   

   product_retail_price product_department                       product_sku  \
0                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   
1                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   
2                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   
3                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   
4                  6.99              Women  2A3E953A5E3D81E67945BCE5519F84C8   

   product_distribution_center_id  
0                               7  
1                               7  
2                               7  
3                               7  
4                               7

In [171]:
# dropping product_sku column
df_inventory_items = df_inventory_items.drop(columns = ['product_sku'])

In [16]:
df_inventory_items.describe()

id     product_id           cost  product_retail_price  \
count  490705.000000  490705.000000  490705.000000         490705.000000   
mean   245353.000000   15254.164465      28.658986             59.569440   
std    141654.476259    8410.586173      31.009435             66.592452   
min         1.000000       1.000000       0.008300              0.020000   
25%    122677.000000    7984.000000      11.368000             24.500000   
50%    245353.000000   15970.000000      19.832000             39.990002   
75%    368029.000000   22502.000000      34.550301             69.949997   
max    490705.000000   29120.000000     557.151002            999.000000   

       product_distribution_center_id  
count                   490705.000000  
mean                         5.011514  
std                          2.900016  
min                          1.000000  
25%                          2.000000  
50%                          5.000000  
75%                          8.000000  
max                         10.000000

In [17]:
# checking missing values
df_inventory_items.isnull().sum()

id                                     0
product_id                             0
created_at                             0
sold_at                           308946
cost                                   0
product_category                       0
product_name                          29
product_brand                        401
product_retail_price                   0
product_department                     0
product_sku                            0
product_distribution_center_id         0
dtype: int64

In [18]:
# Create a brand map to pick the most frequent product_name used according to product_id that is not null
brand_map_product = (
    df_inventory_items[df_inventory_items['product_name'].notnull()]
    .groupby('product_id')['product_name']
    .agg(lambda x: x.mode().iloc[0])
)

In [22]:
# Fill missing product_name using the brand_map_product
df_inventory_items['product_name'] = df_inventory_items.apply(
    lambda row: brand_map_product[row['product_id']] if pd.isnull(row['product_name']) and row['product_id'] in brand_map_product else row['product_name'],
    axis=1
)

In [23]:
# checking missing values
df_inventory_items.isnull().sum()

id                                     0
product_id                             0
created_at                             0
sold_at                           308946
cost                                   0
product_category                       0
product_name                          29
product_brand                        401
product_retail_price                   0
product_department                     0
product_sku                            0
product_distribution_center_id         0
dtype: int64

In [24]:
# Create a brand map to pick the most frequent product_brand used according to product_name that is not null
brand_map_brand = (
    df_inventory_items[df_inventory_items['product_brand'].notnull()]
    .groupby('product_name')['product_brand']
    .agg(lambda x: x.mode().iloc[0])
)

In [27]:
# Fill missing product_name using the brand_map_brand
df_inventory_items['product_brand'] = df_inventory_items.apply(
    lambda row: brand_map_brand[row['product_name']] if pd.isnull(row['product_brand']) and row['product_name'] in brand_map_brand else row['product_brand'],
    axis=1
)

In [28]:
# checking missing values
df_inventory_items.isnull().sum()

id                                     0
product_id                             0
created_at                             0
sold_at                           308946
cost                                   0
product_category                       0
product_name                          29
product_brand                        365
product_retail_price                   0
product_department                     0
product_sku                            0
product_distribution_center_id         0
dtype: int64

_Previous Missing values in 'product_brand' = 401
New Missing values in 'product_brand' = 365_

In [118]:
df_inventory_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 490705 entries, 0 to 490704
Data columns (total 12 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              490705 non-null  int64  
 1   product_id                      490705 non-null  int64  
 2   created_at                      490705 non-null  object 
 3   sold_at                         181759 non-null  object 
 4   cost                            490705 non-null  float64
 5   product_category                490705 non-null  object 
 6   product_name                    490676 non-null  object 
 7   product_brand                   490304 non-null  object 
 8   product_retail_price            490705 non-null  float64
 9   product_department              490705 non-null  object 
 10  product_sku                     490705 non-null  object 
 11  product_distribution_center_id  490705 non-null  int64  
dtypes: float64(2), i

### Data cleaning: Order Items

In [29]:
df_order_items.head()

id  order_id  user_id  product_id  inventory_item_id     status  \
0  152013    104663    83582       14235             410368  Cancelled   
1   40993     28204    22551       14235             110590   Complete   
2   51224     35223    28215       14235             138236   Complete   
3   36717     25278    20165       14235              99072    Shipped   
4  131061     90241    71954       14235             353798    Shipped   

                  created_at                 shipped_at  \
0  2023-05-07 06:08:40+00:00                        NaN   
1  2023-03-14 03:47:21+00:00  2023-03-15 22:57:00+00:00   
2  2023-12-05 13:25:30+00:00  2023-12-06 01:20:00+00:00   
3  2023-12-22 20:48:19+00:00  2023-12-24 16:44:00+00:00   
4  2022-06-19 16:57:59+00:00  2022-06-19 19:29:00+00:00   

                delivered_at returned_at  sale_price  
0                        NaN         NaN        0.02  
1  2023-03-18 01:08:00+00:00         NaN        0.02  
2  2023-12-10 10:04:00+00:00         NaN        0.02  
3                        NaN         NaN        0.02  
4                        NaN         NaN        0.02

In [169]:
df_order_items['order_id'].value_counts

<bound method IndexOpsMixin.value_counts of 0         104663
1          28204
2          35223
3          25278
4          90241
           ...  
181754      6679
181755      5416
181756     43364
181757     73418
181758     98984
Name: order_id, Length: 181759, dtype: int64>

In [30]:
df_order_items.describe()

id       order_id        user_id     product_id  \
count  181759.000000  181759.000000  181759.000000  181759.000000   
mean    90880.000000   62579.982609   49983.272020   15254.151838   
std     52469.448126   36125.987285   28830.895739    8413.482833   
min         1.000000       1.000000       1.000000       1.000000   
25%     45440.500000   31266.500000   25031.500000    7979.500000   
50%     90880.000000   62529.000000   49983.000000   15966.000000   
75%    136319.500000   93845.500000   74867.500000   22504.000000   
max    181759.000000  125226.000000  100000.000000   29120.000000   

       inventory_item_id     sale_price  
count      181759.000000  181759.000000  
mean       245291.860007      59.568544  
std        141661.912894      66.657262  
min             3.000000       0.020000  
25%        122565.000000      24.500000  
50%        245284.000000      39.990002  
75%        367983.000000      69.949997  
max        490705.000000     999.000000

In [31]:
# Checking for consistency
df_order_items['status'].value_counts()

status
Shipped       54440
Complete      45609
Processing    36388
Cancelled     27090
Returned      18232
Name: count, dtype: int64

In [32]:
# checking for duplicates
df_order_items[df_order_items.duplicated()]

Empty DataFrame
Columns: [id, order_id, user_id, product_id, inventory_item_id, status, created_at, shipped_at, delivered_at, returned_at, sale_price]
Index: []

In [33]:
# checking for missing values
df_order_items.isnull().sum()

id                        0
order_id                  0
user_id                   0
product_id                0
inventory_item_id         0
status                    0
created_at                0
shipped_at            63478
delivered_at         117918
returned_at          163527
sale_price                0
dtype: int64

_Nothing important to handle_

In [119]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181759 entries, 0 to 181758
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   id                 181759 non-null  int64  
 1   order_id           181759 non-null  int64  
 2   user_id            181759 non-null  int64  
 3   product_id         181759 non-null  int64  
 4   inventory_item_id  181759 non-null  int64  
 5   status             181759 non-null  object 
 6   created_at         181759 non-null  object 
 7   shipped_at         118281 non-null  object 
 8   delivered_at       63841 non-null   object 
 9   returned_at        18232 non-null   object 
 10  sale_price         181759 non-null  float64
dtypes: float64(1), int64(5), object(5)
memory usage: 15.3+ MB


### Data cleaning: Events

In [34]:
df_events.head()

id  user_id  sequence_number                            session_id  \
0  2198523      NaN                3  83889ed2-2adc-4b9a-af5d-154f6998e778   
1  1773216      NaN                3  7a3fc3f2-e84f-44fe-8876-eff76741f7a3   
2  2380515      NaN                3  13d9b2fb-eee1-43fd-965c-267b38dd7125   
3  2250597      NaN                3  96f1d44e-9621-463c-954c-d8deb7fffe7f   
4  1834446      NaN                3  d09dce10-a7cb-47d3-a9af-44975566fa03   

                  created_at       ip_address          city      state  \
0  2021-06-17 17:30:00+00:00    138.143.9.202     São Paulo  São Paulo   
1  2020-08-07 08:41:00+00:00    85.114.141.79  Santa Isabel  São Paulo   
2  2021-02-15 18:48:00+00:00  169.250.255.132     Mairiporã  São Paulo   
3  2022-03-30 10:56:00+00:00   137.25.222.160       Cajamar  São Paulo   
4  2019-09-05 01:18:00+00:00    161.114.4.174     São Paulo  São Paulo   

  postal_code browser traffic_source      uri event_type  
0   02675-031  Chrome        Adwords  /cancel     cancel  
1   07500-000  Safari        Adwords  /cancel     cancel  
2   07600-000      IE        Adwords  /cancel     cancel  
3   07750-000  Chrome        Adwords  /cancel     cancel  
4   09581-680  Chrome          Email  /cancel     cancel

In [127]:
df_events = df_events.drop(columns =['sequence_number', 'session_id', 'ip_address'])

In [129]:
df_events.isnull().sum()

id                      0
user_id           1125671
created_at              0
city                23080
state                   0
postal_code             0
browser                 0
traffic_source          0
uri                     0
event_type              0
dtype: int64

In [130]:
# Create a map to pick the most frequent city used according to postal code that is not null
map_events_city = (
    df_events[df_events['city'].notnull()]
    .groupby('postal_code')['city']
    .agg(lambda x: x.mode().iloc[0])
)

In [131]:
# Fill missing product_name using the brand_map_brand
df_events['city'] = df_events.apply(
    lambda row: map_events_city[row['postal_code']] if pd.isnull(row['city']) and row['postal_code'] in map_events_city else row['city'],
    axis=1
)

In [132]:
df_events.isnull().sum()

id                      0
user_id           1125671
created_at              0
city                22851
state                   0
postal_code             0
browser                 0
traffic_source          0
uri                     0
event_type              0
dtype: int64

_Some cities were updated based on the postal codes others need to be updated manually based on data dictionary_

In [133]:
# finding distinct postal codes that have missing cities
missing_events_postal = df_events[df_events['city'].isnull()]['postal_code'].dropna().unique()

In [134]:
missing_events_postal

array(['30835', '32826', '50016', '65393-000', '65415-000', '69800-000',
       '75370-000', '88201', '20112', '22556', '22602', '28947', '300384',
       '30093', '32828', '33129', '34482', '39260-000', '46394',
       '47350-000', '48044', '48760-000', '48790-000', '55730-000',
       '57360-000', '630-850', '630-857', '631-820', '631-830', '631-840',
       '63560-000', '65278-000', '65380-000', '68230-000', '68540-000',
       '68620-000', '68695-000', '68890-000', '68924-000', '69200-000',
       '69250-000', '69460-000', '69600-000', '69630-000', '69980-000',
       '72016-190', '72025-065', '73900-000', '76868-000', '76980-000',
       '77389', '78850-000', '84600-000', '98926', '20171', '22408',
       '28524', '30016', '30102', '47240-000', '56180-000', '56380-000',
       '630-500', '63610-000', '641-870', '65530-000', '65690-000',
       '68300-000', '68473-000', '68660-000', '69230-000', '69280-000',
       '69735-000', '69880-000', '72110-600', '630-490', '23236', '29142',

In [143]:
# creating a data dictionary combining postal codes with cities
postcode_events_to_city = {
'30835': 'Augusta',
    '32826': 'Orlando',
    '50016': 'Ames',
    '88201': 'Roswell',
    '20112': 'Manassas',
    '22556': 'Stafford',
    '22602': 'Winchester',
    '28947': 'Andrews',
    '30093': 'Norcross',
    '32828': 'Orlando',
    '33129': 'Miami',
    '34482': 'Ocala',
    '46394': 'Whiting',
    '48044': 'Macomb',
    '48760-000': 'Carangola',
    '48790-000': 'Caratinga',
    '55730-000': 'Paracatu',
    '57360-000': 'Palmeira dos Índios',
    '630-850': 'Sapporo',
    '630-857': 'Sapporo',
    '631-820': 'Sapporo',
    '631-830': 'Sapporo',
    '631-840': 'Sapporo',
    '63560-000': 'Iguatu',
    '65278-000': 'Pinheiro',
    '65380-000': 'Barra do Corda',
    '68230-000': 'Almeirim',
    '68540-000': 'Paragominas',
    '68620-000': 'Parauapebas',
    '68695-000': 'Redenção',
    '68890-000': 'Santarém',
    '68924-000': 'Laranjal do Jari',
    '69200-000': 'Itacoatiara',
    '69250-000': 'Manacapuru',
    '69460-000': 'Tefé',
    '69600-000': 'Coari',
    '69630-000': 'Codajás',
    '69980-000': 'Cruzeiro do Sul',
    '72016-190': 'Brasília',
    '72025-065': 'Brasília',
    '73900-000': 'Formosa',
    '76868-000': 'Ariquemes',
    '76980-000': 'Ji-Paraná',
    '77389': 'Spring',
    '78850-000': 'Cáceres',
    '84600-000': 'União da Vitória',
    '98926': 'Ellensburg',
    '20171': 'Herndon',
    '22408': 'Fredericksburg',
    '28524': 'Beaufort',
    '30016': 'Covington',
    '30102': 'Acworth',
    '56180-000': 'Petrolina',
    '56380-000': 'Ouricuri',
    '63610-000': 'Crateús',
    '65530-000': 'Chapadinha',
    '65690-000': 'Barra do Corda',
    '68300-000': 'Altamira',
    '68473-000': 'Tucuruí',
    '68660-000': 'Marabá',
    '69230-000': 'Parintins',
    '69280-000': 'Maués',
    '69735-000': 'Humaitá',
    '69880-000': 'Eirunepé',
    '72110-600': 'Brasília',
    '23236': 'Richmond',
    '29142': 'Vila Velha',
    '30028': 'Cumming',
    '30044': 'Lawrenceville',
    '32092': 'Saint Augustine',
    '46988': 'Huntington',
    '48475-000': 'Itamarandiba',
    '55920-000': 'Ouricuri',
    '62150-000': 'Quixadá',
    '63540-000': 'Iguatu',
    '65365-000': 'Bacabal',
    '68330-000': 'Altamira',
    '69435-000': 'Tefé',
    '44330-000': 'Cruz das Almas',
    '50197': 'Cologne',
    '22406': 'Fredericksburg',
    '23693': 'Yorktown',
    '32940': 'Melbourne',
    '33579': 'Riverview',
    '65725-000': 'Balsas',
    '68250-000': 'Almeirim',
    '68617-000': 'Parauapebas',
    '69190-000': 'Itacoatiara',
    '69830-000': 'Coari',
    '77379': 'Spring',
    '17015': 'York',
    '18302': 'East Stroudsburg',
    '19803': 'Wilmington',
    '23188': 'Williamsburg',
    '68650-000': 'Parauapebas',
    '22407': 'Fredericksburg',
    '56460-000': 'Petrolina',
    '642-370': 'Sapporo',
    '70706': 'Denham Springs',
    '12065': 'Clifton Park',
    '38434': 'Uberlândia',
    '50191': 'Cologne',
    '79706': 'Midland',
    '630-520': 'Sapporo',
    '642-315': 'Sapporo',
    '15174': 'Pittsburgh',
    '32060': 'Live Oak',
    '38297': 'Uberlândia',
    '33437': 'Boynton Beach',
    '630-856': 'Sapporo',
    '641-860': 'Sapporo',
    '25191': 'Charleston',
    '27537': 'Raleigh',
    '28909': 'Andrews',
    '28946': 'Andrews',
    '30162': 'Rome',
    '631-810': 'Sapporo',
    '641-465': 'Sapporo',
    '41130': 'Florence',
    '641-920': 'Sapporo',
    '34476': 'Ocala',
    '50620': 'Cologne',
    '50059': 'Ames',
    '69800-000': 'Humaitá',
    '69800': 'Saint-Priest', 
    '69800': 'Heroica Ciudad de Tlaxiaco',
    '47350-000': 'Panjar',
    '47240-000': 'Gornja Mocila',
    '65393-000': 'Buriticupu',
    '65415-000': 'Coroatá',
    '75370-000': 'Goianira',
    '39260-000': 'Várzea da Palma'
}

In [144]:
# filling missing cities based on the dictionary created.
df_events['city'] = df_events.apply(
    lambda row: postcode_events_to_city.get(row['postal_code'], row['city']) if pd.isnull(row['city']) else row['city'],
    axis=1
)

In [154]:
df_events.isnull().sum()

id                      0
user_id           1125671
created_at              0
city                  815
state                   0
postal_code             0
browser                 0
traffic_source          0
uri                     0
event_type              0
dtype: int64

_Previous missing values: 23,080
Updated Missing values: 815_

In [35]:
df_events.describe()

id       user_id  sequence_number
count  2.431963e+06  1.306292e+06     2.431963e+06
mean   1.215982e+06  4.996636e+04     3.316146e+00
std    7.020474e+05  2.881711e+04     2.676365e+00
min    1.000000e+00  1.000000e+00     1.000000e+00
25%    6.079915e+05  2.500500e+04     1.000000e+00
50%    1.215982e+06  4.992700e+04     2.000000e+00
75%    1.823972e+06  7.483900e+04     4.000000e+00
max    2.431963e+06  1.000000e+05     1.300000e+01

In [36]:
df_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2431963 entries, 0 to 2431962
Data columns (total 13 columns):
 #   Column           Dtype  
---  ------           -----  
 0   id               int64  
 1   user_id          float64
 2   sequence_number  int64  
 3   session_id       object 
 4   created_at       object 
 5   ip_address       object 
 6   city             object 
 7   state            object 
 8   postal_code      object 
 9   browser          object 
 10  traffic_source   object 
 11  uri              object 
 12  event_type       object 
dtypes: float64(1), int64(2), object(10)
memory usage: 241.2+ MB


In [37]:
df_events['event_type'].value_counts()

event_type
product       845607
cart          595994
department    595323
purchase      181759
cancel        125568
home           87712
Name: count, dtype: int64

### Data cleaning: Distribution Centers

In [164]:
df_distribution_centers.head()

id            name  latitude  longitude
0   1      Memphis TN   35.1174   -89.9711
1   2      Chicago IL   41.8369   -87.6847
2   3      Houston TX   29.7604   -95.3698
3   4  Los Angeles CA   34.0500  -118.2500
4   5  New Orleans LA   29.9500   -90.0667

In [121]:
df_distribution_centers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         10 non-null     int64  
 1   name       10 non-null     object 
 2   latitude   10 non-null     float64
 3   longitude  10 non-null     float64
dtypes: float64(2), int64(1), object(1)
memory usage: 452.0+ bytes


# Exporting Final Datasets

In [177]:
# exporting all datasets
df_orders.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_1.0.pkl'))
df_products.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'products_1.0.pkl'))
df_users.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'users_1.0.pkl'))
df_inventory_items.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'inventory_items_1.0.pkl'))
df_order_items.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'order_items_1.0.pkl'))
df_events.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'events_1.0.pkl'))
df_distribution_centers.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'distribution_centers_1.0.pkl'))